### PR TITLE
Implementing federation spec

### PR DESCRIPTION
**Pull request by dkuc84**
Create new `graphql-kotlin-federation` module that extends `graphql-kotlin-schema-generator` to generate federated GraphQL schemas. Federated GraphQL architecture is an alternative to schema stitching approach that relies on declarative programming composition of schemas.

`graphql-kotlin-federation` adds support for the following:
- new `@key`, `@extends`, `@external`, `@requires` and `@provides` directives
- new `_FieldSet` and `_Any` scalar types
- new `_Entity` union type of all federated types
- new `_service` query that returns `_Service` object with `sdl` field
- new `_entity` query that returns list of `_Entity` objects 

In order to generate federate schemas we have to annotate the underlying type with proper new directives and then use new `toFederatedSchema` function to build the schema (instead of `toSchema` function from `graphql-kotlin-schema-generator`). `FederatedSchemaGeneratorHooks` also require to specify `FederatedTypeRegistry` that holds a mapping between federated type names and their corresponding resolver. Those new `FederatedTypeResolver`s are used to instantiate the federated types based on the federated `_entity` query generated from the gateway.

See: https://www.apollographql.com/docs/apollo-server/federation/introduction/

Resolves: #238

### CODE REVIEW

File: graphql-kotlin-federation/pom.xml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    <name>graphql-kotlin-federation</name>
+    <packaging>jar</packaging>
+
+    <build>
+        <sourceDirectory>${project.basedir}/src/main/kotlin</sourceDirectory>
+        <testSourceDirectory>${project.basedir}/src/test/kotlin</testSourceDirectory>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-antrun-plugin</artifactId>
```
Comment (pcarrier) [inline graphql-kotlin-federation/pom.xml]: Does this do something I cannot find?

File: graphql-kotlin-federation/src/main/kotlin/com/expedia/graphql/federation/FederationSchemaGeneratorHooks.kt [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            .asSequence()
+            .filterIsInstance<GraphQLObjectType>()
+            .filter { type -> type.getDirective("key") != null }
+            .onEach { type -> type.validateFederatedType() }
+            .map { it.name }
+            .toSet()
+
+        // register new federated queries
+        if (entityTypeNames.isNotEmpty()) {
+            federatedQuery.field(SERVICE_FIELD_DEFINITION)
```
Comment (pcarrier) [inline graphql-kotlin-federation/src/main/kotlin/com/expedia/graphql/federation/FederationSchemaGeneratorHooks.kt]: A service could be federated without exposing any entities, eg to provide new fields on the query or mutation root types.

File: graphql-kotlin-federation/pom.xml (hunk 1 above)
Comment (dkuc84) [reply graphql-kotlin-federation/pom.xml]: It applies detekt and ktlint plugins from the parent (https://github.com/ExpediaDotCom/graphql-kotlin/blob/master/pom.xml#L138). Cannot apply them at the root project as it doesn't have any source code and there are no official maven plugins that would work out of the box.

File: graphql-kotlin-federation/src/main/kotlin/com/expedia/graphql/federation/FederationSchemaGeneratorHooks.kt (hunk 2 above)
Comment (dkuc84) [reply graphql-kotlin-federation/src/main/kotlin/com/expedia/graphql/federation/FederationSchemaGeneratorHooks.kt]: nice catch! thanks

File: graphql-kotlin-federation/src/test/kotlin/com/expedia/graphql/federation/FederatedSchemaGeneratorTest.kt [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+    @Test
+    fun `verify generator does not add federation queries for non-federated schemas`() {
+        val expectedSchema = """
+            schema {
+              query: Query
+            }
+
+            type Query {
+              _service: _Service
```
Comment (gscheibel) [inline graphql-kotlin-federation/src/test/kotlin/com/expedia/graphql/federation/FederatedSchemaGeneratorTest.kt]: If the service is not "federate-able" should its SDL contain `_service`?

Comment (dkuc84) [reply graphql-kotlin-federation/src/test/kotlin/com/expedia/graphql/federation/FederatedSchemaGeneratorTest.kt]: yes -> https://github.com/ExpediaDotCom/graphql-kotlin/pull/280#discussion_r309385781

File: graphql-kotlin-federation/README.md [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+```groovy
+compile(group: 'com.expedia', name: 'graphql-kotlin-federation', version: "$latestVersion")
+```
+
+## Usage
+
+In order to generate valid federated schemas, you will need to annotate your both base service and the one extending it. Federated Gateway (e.g. Apollo) will then combine the individual graphs to form single federated graph.
+
+#### Base Service
```
[the remark this answers is not in the document]
Comment (dkuc84) [reply graphql-kotlin-federation/README.md]: i don't think that name matters - base service is just a service that defines base type, i'll update it to be more specific

File: graphql-kotlin-federation/src/main/kotlin/com/expedia/graphql/federation/directives/ProvidesDirective.kt [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+ *   @ProvidesDirective(FieldSet("name"))
+ *   fun user(): User = // implementation goes here
+ * }
+ *
+ * @KeyDirective(FieldSet("userId"))
+ * @ExtendsDirective
+ * class User(
+ *   @property:ExternalDirective val userId: String,
+ *   @property:ExternalDirective val name: String
+ * )
```
[the remark this answers is not in the document]
Comment (dkuc84) [reply graphql-kotlin-federation/src/main/kotlin/com/expedia/graphql/federation/directives/ProvidesDirective.kt]: it would result in invalid schema

**Pull request merged**